File: Rogue-Gos/Daily Coding Problem/From Book/Liniked Lists/linked_list.py
class Node:
    def __init__(self, data, next=None, prev=None):
        self.data = data
        # Single Link going forward
        self.next = next
        # Double link
        self.prev = prev

    def reverse(node):
        #  _reverse() reverses and returns both head and tail.
        # aConventionally, an underscore denotes an unused variable.
        head, _ = _reverse(node)
        return head

    def _reverse(node):
        if node is None:
            return None, None
        if node.next is None:
            return node, node

    # Reverse rest of linked list and move node to after tail.
        head, tail = _reverse(node.next)
        node.next = None
        tail.next = node
        return head, node

    def reverse(head):
        prev, current = None, head
        while current is not None:
            # Make current node point to prev and move both forward one
            tmp = current.next
            current.next = prev
            prev = current
            current = tmp
        return prev

File: Rogue-Gos/Daily Coding Problem/From Book/Liniked Lists/linkedList.js
/*
create Singly Linked List Class and Node Class

*/

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }
    // determine if empty
    isEmpty() {
        return this.head == null
    }
    // add to back
    addToBack(data) {
        let newNode = new Node(data);
        if(this.isEmpty()) {
            this.head = newNode;
            return this
        }
        let runner = this.head;
        while(runner.next != null) {
            runner = runner.next;
        }        
        runner.next = newNode;
        return this
    }
    // display list
    displayList() {
        if(this.isEmpty()) {
            return null
        }
        let runner = this.head;
        while(runner.next != null) {
            console.log(runner)
            runner = runner.next;
        }
    }
    // reverse
}

let myList = new LinkedList();
myList.addToBack(1).addToBack(2).addToBack(3).displayList();
myList.addToBack(3);
myList.addToBack(6);
myList.addToBack(2).displayList();